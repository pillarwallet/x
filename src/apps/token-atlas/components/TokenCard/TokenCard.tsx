import { useState } from 'react';

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import Body from '../Typography/Body';

type TokenCardProps = {
  tokenLogo?: string;
  tokenName?: string;
  tokenSymbol?: string;
  blockchainLogo?: string;
  onClick: () => void;
};

const TokenCard = ({
  tokenLogo,
  tokenName,
  tokenSymbol,
  blockchainLogo,
  onClick,
}: TokenCardProps) => {
  const [isBrokenImage, setIsBrokenImage] = useState<boolean>(false);

  return (
    <div
      id="token-atlas-token-card"
      className="flex flex-col relative w-[108px] h-[125px] bg-medium_grey rounded-lg px-4 pb-4 pt-6 items-center justify-center cursor-pointer"
      onClick={onClick}
      data-testid="token-card"
    >
      {blockchainLogo && (
        <img
          src={blockchainLogo}
          alt="chain-logo"
          className="absolute top-2 right-2 w-4 h-4 object-fill rounded-full"
          data-testid="token-card-chain-logo"
        />
      )}

      <div className="relative w-[40px] h-[40px] rounded-full">
        {tokenLogo && !isBrokenImage ? (
          <img
            src={tokenLogo}
            alt="token-logo"
            className="w-full h-full object-fill rounded-full"
            data-testid="token-card-token-logo"
            onError={() => setIsBrokenImage(true)}
          />
        ) : (
          <div className="w-full h-full overflow-hidden rounded-full">
            <RandomAvatar name={`${tokenName}-chain` || ''} />
          </div>
        )}

        {/* Overlay text when no token logo available */}
        {(!tokenLogo || isBrokenImage) && (
          <span className="absolute inset-0 flex items-center justify-center text-white text-base font-bold">
            {tokenName?.slice(0, 2)}
          </span>
        )}
      </div>

      <Body className="text-base capitalize w-full truncate text-center">
        {tokenName}
      </Body>
      <Body className="text-white_grey">{tokenSymbol}</Body>
    </div>
  );
};

export default TokenCard;
