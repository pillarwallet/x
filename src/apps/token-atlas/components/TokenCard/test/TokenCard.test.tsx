/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

// components
import TokenCard from '../TokenCard';

describe('<TokenCard />', () => {
  const mockOnClick = jest.fn();

  const defaultProps = {
    tokenLogo: 'token-logo.png',
    tokenName: 'Test Token',
    tokenSymbol: 'TTK',
    blockchainLogo: 'blockchain-logo.png',
    onClick: mockOnClick,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer.create(<TokenCard {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with all props', () => {
    render(<TokenCard {...defaultProps} />);

    const tokenLogo = screen.getByTestId('token-card-token-logo');
    expect(tokenLogo).toHaveAttribute('src', 'token-logo.png');

    const blockchainLogo = screen.getByTestId('token-card-chain-logo');
    expect(blockchainLogo).toHaveAttribute('src', 'blockchain-logo.png');

    expect(screen.getByText('Test Token')).toBeInTheDocument();
    expect(screen.getByText('TTK')).toBeInTheDocument();
  });

  it('renders with default tokenLogo and default blockchainLogo (random avatar)', () => {
    const propsWithoutLogos = {
      tokenName: undefined,
      tokenSymbol: undefined,
      onClick: mockOnClick,
    };

    render(<TokenCard {...propsWithoutLogos} />);

    expect(
      screen.queryByTestId('token-card-token-logo')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('token-card-chain-logo')
    ).not.toBeInTheDocument();
    expect(screen.queryByText('Test Token')).not.toBeInTheDocument();
    expect(screen.queryByText('TTK')).not.toBeInTheDocument();
    expect(screen.getAllByTestId('random-avatar')).toHaveLength(1);
  });

  it('calls onClick when clicked', () => {
    render(<TokenCard {...defaultProps} />);

    const tokenCard = screen.getByTestId('token-card');
    fireEvent.click(tokenCard);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });
});
