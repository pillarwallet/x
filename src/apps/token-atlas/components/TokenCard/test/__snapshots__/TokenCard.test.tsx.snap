// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenCard /> renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col relative w-[108px] h-[125px] bg-medium_grey rounded-lg px-4 pb-4 pt-6 items-center justify-center cursor-pointer"
  data-testid="token-card"
  id="token-atlas-token-card"
  onClick={[MockFunction]}
>
  <img
    alt="chain-logo"
    className="absolute top-2 right-2 w-4 h-4 object-fill rounded-full"
    data-testid="token-card-chain-logo"
    src="blockchain-logo.png"
  />
  <div
    className="relative w-[40px] h-[40px] rounded-full"
  >
    <img
      alt="token-logo"
      className="w-full h-full object-fill rounded-full"
      data-testid="token-card-token-logo"
      src="token-logo.png"
    />
  </div>
  <p
    className="text-[13px] font-semibold text-base capitalize w-full truncate text-center"
  >
    Test Token
  </p>
  <p
    className="text-[13px] font-semibold text-white_grey"
  >
    TTK
  </p>
</div>
`;
